sync: simplify and remove unused code from sync.h

vasild opened the pull request:
Summary:

* Reduce 4 of the `MaybeCheckNotHeld()` definitions to 2 by using a template.
* Remove unused template parameter from `::UniqueLock`.
* Use `MutexType` instead of `Mutex` for a template parameter name to avoid overlap/confusion with the `Mutex` class.
* Rename `AnnotatedMixin::UniqueLock` to `AnnotatedMixin::unique_lock` to avoid overlap/confusion with the global `UniqueLock` and for consistency with `UniqueLock::reverse_lock`.

The first commit `sync: simplify MaybeCheckNotHeld() definitions by using a template` is also part of https://github.com/bitcoin/bitcoin/pull/25390

vasild committed:
sync: simplify MaybeCheckNotHeld() definitions by using a template

Reduce 4 of the `MaybeCheckNotHeld()` definitions to 2 by using a
template. This also makes the function usable for other
[BasicLockable](https://en.cppreference.com/w/cpp/named_req/BasicLockable)
types.

vasild committed:
sync: remove unused template parameter from ::UniqueLock

The template parameter `typename Base = typename Mutex::UniqueLock` is
not used, so remove it. Use internally defined type `Base` to avoid
repetitions of `Mutex::UniqueLock`.

vasild committed:
sync: avoid confusing name overlap (Mutex)

Use `MutexType` instead of `Mutex` for the template parameter of
`UniqueLock` because there is already a class named `Mutex` and the
naming overlap is confusing. `MutexType` is used elsewhere in `sync.h`.

vasild committed:
sync: remove DebugLock alias template

Use `UniqueLock` directly. Type deduction works just fine from the first
argument to the constructor of `UniqueLock`, so there is no need to
repeat

```cpp
UniqueLock<typename std::remove_reference<typename std::remove_pointer<decltype(cs)>::type>::type>
```

five times in the `LOCK` macros. Just `UniqueLock` suffices.

MarcoFalke commented on src/sync.h:
```diff
@@ -148,72 +148,74 @@ inline void AssertLockNotHeldInline(const char* name, const char* file, int line
+class SCOPED_LOCKABLE DebugLock : public MutexType::UniqueLock
 {
 private:
+    using PARENT = typename MutexType::UniqueLock;
```
I don't think we need to unify the names. Base seems fine to keep as is

vasild replied:
Changed to `Base`.

vasild commented:
`332cfac809...19fce0321b`: address suggestion

vasild commented:
`19fce0321b...f5b529369c`: rebase due to conflict

ryanofsky reviewed (approved):
Code review ACK f5b529369c8e5dee655ed559fd9583947188b18c. This makes nice simplifications and gets rid of some confusing shadowing, but I would suggest an alternative to the last commit.

I don't think the last commit "sync: rename the global `UniqueLock` to `DebugLock`" (f5b529369c8e5dee655ed559fd9583947188b18c) is good because it gets rid of the naming consistency between standard c++ thread primitives and our thread primitives. The standard c++ mutex class is called `std::mutex` while ours is called `Mutex`. The standard c++ recursive mutex class is called `std::recursive_mutex` while ours is called `RecursiveMutex`. Than standard c++ lock class is called `std::unique_lock`, so I think ours should be called `UniqueLock`, not `DebugLock`. It is true that the main features our lock class adds currently are related to debugging (though it also adds support for reverse locking), but we are not naming our custom classes after what features they add, otherwise we would be calling `Mutex` `AnnotatedMutex` etc. And I don't think it is a good idea to name our locking primitives after the particular features they add, because we should have flexibility to add and remove features from our locking primitives in the future.

Another benefit of dropping the last commit is that it would confine changes in this PR to `sync.h` and not change outside code in `src/node/interfaces.cpp`.

I do think one benefit of the last commit is that disambiguates global `UniqueLock` from `AnnotatedMixin::UniqueLock`. But I would fix this by renaming `AnnotatedMixin::UniqueLock` to `AnnotatedMixin::unique_lock`. I think `AnnotatedMixin::unique_lock` name makes sense because it is a typedef for `std::unique_lock`, and would also be consistent with current `UniqueLock::reverse_lock`.

vasild committed:
sync: rename AnnotatedMixin::UniqueLock to AnnotatedMixin::unique_lock

This avoids confusion with the global `UniqueLock` and the snake case
is consistent with `UniqueLock::reverse_lock.

vasild commented on src/sync.h:
```diff
@@ -148,11 +148,11 @@ inline void AssertLockNotHeldInline(const char* name, const char* file, int line
 #define AssertLockNotHeld(cs) AssertLockNotHeldInline(#cs, __FILE__, __LINE__, &cs)
 
 /** Wrapper around std::unique_lock style lock for Mutex. */
```
Done.

vasild commented:
`f5b529369c...75c3f9f880`: @ryanofsky, done!

fanquake commented:
> done!

@vasild looks like you now also need to update the PR description?

vasild commented:
@fanquake, updated, thanks!

aureleoules reviewed (approved):
ACK 75c3f9f8806259ac7ac02e725d2f2f48e5a1d954 - LGTM
I verified all commits compile and unit tests pass.

ryanofsky reviewed (approved):
Code review ACK 75c3f9f8806259ac7ac02e725d2f2f48e5a1d954. Nice cleanups! Just suggested changes since last review: keeping UniqueLock name and fixing a missed rename in a code comment

ajtowns commented:
Post-merge ACK 75c3f9f8806259ac7ac02e725d2f2f48e5a1d954

Nice cleanup, especially with the final simplifications

jonatack commented:
Post-merge ACK 75c3f9f8806259ac7ac02e725d2f2f48e5a1d954

hebasto reviewed (commented):
Post-merge ACK 75c3f9f8806259ac7ac02e725d2f2f48e5a1d954, I have reviewed the code and it looks OK.
